controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,246 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/utils/strings/slices"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var ControllerName = "constraintstatus_reconciler"
+
+type ConstraintPodStatusReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	Log           logr.Logger
+	DynamicClient dynamic.Interface
+	Namespace     string
+	// This includes api-resources list and it finds a missing version of resources.
+	DiscoveryStorage *DiscoveryStorage
+	// key = constraintPodName
+	ConstraintToSyncOnly map[string][]v1alpha1.SyncOnlyEntry
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ConstraintPodStatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{MaxConcurrentReconciles: int(1)}).
+		Named(ControllerName).
+		For(&v1beta1.ConstraintPodStatus{},
+			builder.WithPredicates(predicate.Funcs{
+				// Execute this reconcile func when it is audit-constraintStatuspod
+				// because a constraint creates 4 constraintPodstatus
+				CreateFunc: func(e event.CreateEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					oldObj := e.ObjectOld.(*v1beta1.ConstraintPodStatus)
+					newObj := e.ObjectNew.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(newObj.Status.Operations, "audit") &&
+						// Update when the constraint is refreshed
+						oldObj.Status.ObservedGeneration != newObj.Status.ObservedGeneration
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+			},
+			)).
+		Complete(r)
+}
+
+// User set gatekeeper.spec.audit.auditFromCache to Automatic, this reconcile function
+// collect requested constraints and add found kinds which is used in constraint to config.spec.sync.syncOnly
+func (r *ConstraintPodStatusReconciler) Reconcile(ctx context.Context,
+	request reconcile.Request,
+) (reconcile.Result, error) {
+	log := r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling ConstraintPodStatus and Config")
+	// This is used for RequeueAfter
+	var requeueTime time.Duration
+
+	gatekeeper := &operatorv1alpha1.Gatekeeper{}
+	// Get gatekeeper resource
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: "",
+		Name:      "gatekeeper",
+	}, gatekeeper)
+	if err != nil {
```

Comment:
You should guard against not found. When it's not found, you shouldn't retry.

Logging the error is probably useful here too.

Reply:
event though this controller starts througn gatekeeper resource?

Reply:
Yeah, just in case there is a gap in time between when this manager/controller stops, we don't want to have an infinite loop (granted it does have an exponential backoff).